File: src/personal_data_warehouse_voice_memos/store_writer.py
"""Core Data writer for the Voice Memos CloudRecordings.db store.

Renames must be real Core Data saves, not raw SQL: the store belongs to
NSPersistentCloudKitContainer (mirrored by the voicememod daemon), and only a
save that records persistent history gets exported to CloudKit and synced to
the user's other devices. A direct sqlite UPDATE would stay local-only and be
clobbered by the next CloudKit import.

Two details make this safe against a store we do not own:

- The managed object model is loaded from the store's own Z_MODELCACHE table
  (the exact model the daemon last saved with), never from a bundled copy, so
  OS updates are tracked automatically.
- The store is opened with automatic migration disabled: if Apple ever ships
  an incompatible model, opening fails loudly instead of mutating the store.
"""

from __future__ import annotations

from pathlib import Path
import sqlite3
from typing import Sequence
import zlib

from personal_data_warehouse_voice_memos.writeback import (
    AUTO_NAMED_FLAG,
    RenamePlanItem,
    TRANSACTION_AUTHOR,
    is_auto_named,
)

STATUS_RENAMED = "renamed"
STATUS_WOULD_RENAME = "would_rename"
STATUS_SKIPPED_NOT_AUTO_NAMED = "skipped_not_auto_named"
STATUS_SKIPPED_MISSING = "skipped_missing"


def load_cached_model_data(store_path: Path | str) -> bytes:
    """Read the archived managed object model cached inside the store.

    Core Data stores it as a raw-DEFLATE-compressed keyed archive in
    Z_MODELCACHE (Apple's COMPRESSION_ZLIB is raw deflate, no zlib header).
    """
    connection = sqlite3.connect(f"file:{Path(store_path)}?mode=ro", uri=True)
    try:
        row = connection.execute("SELECT Z_CONTENT FROM Z_MODELCACHE LIMIT 1").fetchone()
    finally:
        connection.close()
    if row is None or row[0] is None:
        raise RuntimeError(f"store has no cached managed object model: {store_path}")
    return zlib.decompress(row[0], -15)


def _load_cached_model(store_path: Path | str):
    from Foundation import NSData, NSKeyedUnarchiver

    plist = load_cached_model_data(store_path)
    data = NSData.dataWithBytes_length_(plist, len(plist))
    unarchiver, error = NSKeyedUnarchiver.alloc().initForReadingFromData_error_(data, None)
    if unarchiver is None:
        raise RuntimeError(f"could not decode cached model archive: {error}")
    unarchiver.setRequiresSecureCoding_(False)
    model = unarchiver.decodeObjectForKey_("root")
    if model is None:
        raise RuntimeError("cached model archive did not contain a managed object model")
    return model


def _open_store(store_path: Path | str, model):
    from CoreData import (
        NSPersistentContainer,
        NSPersistentHistoryTrackingKey,
        NSPersistentStoreDescription,
        NSPersistentStoreRemoteChangeNotificationPostOptionKey,
    )
    from Foundation import NSURL

    description = NSPersistentStoreDescription.persistentStoreDescriptionWithURL_(
        NSURL.fileURLWithPath_(str(store_path))
    )
    # Never migrate someone else's store: an incompatible model must fail
    # loudly, not rewrite the store's schema metadata.
    description.setShouldMigrateStoreAutomatically_(False)
    description.setShouldInferMappingModelAutomatically_(False)
    description.setShouldAddStoreAsynchronously_(False)
    # Record persistent history for our save (what CloudKit mirroring
    # exports) and post the cross-process change notification the daemon and
    # app listen for.
    description.setOption_forKey_(True, NSPersistentHistoryTrackingKey)
    description.setOption_forKey_(True, NSPersistentStoreRemoteChangeNotificationPostOptionKey)

    container = NSPersistentContainer.alloc().initWithName_managedObjectModel_("VoiceMemos", model)
    container.setPersistentStoreDescriptions_([description])
    load_result: dict[str, object] = {}

    def completion(_description, error) -> None:
        load_result["error"] = error

    container.loadPersistentStoresWithCompletionHandler_(completion)
    error = load_result.get("error")
    if error is not None:
        raise RuntimeError(f"could not open Voice Memos store: {error}")
    return container


def apply_renames(
    store_path: Path | str,
    items: Sequence[RenamePlanItem],
    *,
    author: str = TRANSACTION_AUTHOR,
    dry_run: bool = False,
) -> list[dict[str, str]]:
    """Apply planned renames through a single Core Data save.

    Every item is re-checked against the live row before mutating (the memo
    could have been hand-renamed on another device between planning and
    writing); anything no longer auto-named is skipped, never overwritten.
    """
    try:
        from CoreData import NSFetchRequest, NSMergeByPropertyObjectTrumpMergePolicy
        from Foundation import NSPredicate
    except ImportError as exc:  # pragma: no cover - exercised only off-macOS
        raise RuntimeError(
            "Voice Memos write-back requires pyobjc-framework-CoreData (macOS only)"
        ) from exc

    model = _load_cached_model(store_path)
    container = _open_store(store_path, model)
    context = container.viewContext()
    context.setTransactionAuthor_(author)
    context.setName_("pdw-voice-memos-writeback")
    context.setMergePolicy_(NSMergeByPropertyObjectTrumpMergePolicy)

    results: list[dict[str, str]] = []
    changed = False
    for item in items:
        request = NSFetchRequest.fetchRequestWithEntityName_("CloudRecording")
        request.setPredicate_(
            NSPredicate.predicateWithFormat_argumentArray_("uniqueID == %@", [item.unique_id])
        )
        matches, error = context.executeFetchRequest_error_(request, None)
        if matches is None:
            raise RuntimeError(f"fetch failed for {item.unique_id}: {error}")
        if len(matches) == 0:
            results.append({"unique_id": item.unique_id, "status": STATUS_SKIPPED_MISSING})
            continue
        recording = matches[0]
        current_title = str(recording.valueForKey_("encryptedTitle") or "")
        current_flags = int(recording.valueForKey_("flags") or 0)
        if not is_auto_named(current_title, current_flags):
            results.append({"unique_id": item.unique_id, "status": STATUS_SKIPPED_NOT_AUTO_NAMED})
            continue
        if dry_run:
            results.append({"unique_id": item.unique_id, "status": STATUS_WOULD_RENAME})
            continue
        recording.setValue_forKey_(item.new_title, "encryptedTitle")
        recording.setValue_forKey_(item.new_title, "customLabelForSorting")
        recording.setValue_forKey_(current_flags & ~AUTO_NAMED_FLAG, "flags")
        results.append({"unique_id": item.unique_id, "status": STATUS_RENAMED})
        changed = True

    if changed:
        ok, error = context.save_(None)
        if not ok:
            raise RuntimeError(f"Voice Memos store save failed: {error}")
    return results

File: app/internal/uploaders/voicememos/storewriter.go
package voicememos

import (
	"bytes"
	"compress/flate"
	"crypto/sha256"
	_ "embed"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"syscall"
	"time"

	"github.com/zachlatta/personal-data-warehouse/app/internal/uploaders/common"
)

// hostOS is runtime.GOOS; a package var so the macOS-only paths are testable
// on the Linux CI runner.
var hostOS = runtime.GOOS

// Renames must be real Core Data saves, not raw SQL: CloudRecordings.db
// belongs to NSPersistentCloudKitContainer (mirrored by voicememod), and only
// a save that records persistent history is exported to CloudKit and synced
// to the other devices. A direct sqlite UPDATE would stay local and be
// clobbered by the next import.
//
// Go cannot drive Core Data, so the save is done by a small Swift helper
// (macos/VoiceMemosWriteback.swift, embedded below) compiled on demand with
// swiftc and cached under Application Support, rebuilt only when the embedded
// source changes -- the same arrangement the photos exporter uses for its
// PhotoKit helper. The helper does exactly what the Python store_writer did:
// model from the store's own Z_MODELCACHE, migration disabled, persistent
// history tracking + remote change notification, transaction author
// TransactionAuthor, every row re-checked as still auto-named, one save.

// Status values the writer reports per plan item.
const (
	StatusRenamed             = "renamed"
	StatusWouldRename         = "would_rename"
	StatusSkippedNotAutoNamed = "skipped_not_auto_named"
	StatusSkippedMissing      = "skipped_missing"
)

//go:embed macos/VoiceMemosWriteback.swift
var writebackHelperSource []byte

const (
	helperExecutableName    = "voice-memos-writeback"
	helperBuildTimeout      = 10 * time.Minute
	helperRunTimeout        = 5 * time.Minute
	helperDirectoryName     = "voice-memos-writeback"
	helperSourceStampName   = "source.sha256"
	helperBuildLockFileName = "build.lock"
)

// LoadCachedModelData reads the archived managed object model cached inside
// the store. Core Data stores it as a raw-DEFLATE-compressed keyed archive in
// Z_MODELCACHE (Apple's COMPRESSION_ZLIB is raw deflate, no zlib header).
func LoadCachedModelData(storePath string) ([]byte, error) {
	db, err := common.OpenSQLiteReadOnly(storePath)
	if err != nil {
		return nil, err
	}
	defer db.Close()
	rows, err := common.Query(db, "SELECT Z_CONTENT FROM Z_MODELCACHE LIMIT 1")
	if err != nil {
		return nil, err
	}
	if len(rows) == 0 || rows[0].Get("Z_CONTENT") == nil {
		return nil, fmt.Errorf("store has no cached managed object model: %s", storePath)
	}
	blob := rows[0].Bytes("Z_CONTENT")
	if blob == nil {
		return nil, fmt.Errorf("store has no cached managed object model: %s", storePath)
	}
	reader := flate.NewReader(bytes.NewReader(blob))
	defer reader.Close()
	data, err := io.ReadAll(reader)
	if err != nil {
		return nil, fmt.Errorf("could not decode cached model archive: %w", err)
	}
	return data, nil
}

// HelperSourceSHA256 is the embedded helper source's digest, the cache key
// for the compiled binary.
func HelperSourceSHA256() string {
	sum := sha256.Sum256(writebackHelperSource)
	return hex.EncodeToString(sum[:])
}

// DefaultHelperDir is where the compiled helper lives:
// ~/Library/Application Support/personal-data-warehouse/voice-memos-writeback.
func DefaultHelperDir() string {
	return filepath.Join(common.ApplicationSupportDir(), helperDirectoryName)
}

// HelperBinaryPath is the compiled helper inside root.
func HelperBinaryPath(root string) string {
	return filepath.Join(root, helperExecutableName)
}

// HelperIsCurrent reports whether root holds a binary built from the embedded
// source (the stamp file matches the source digest).
func HelperIsCurrent(root string) bool {
	info, err := os.Stat(HelperBinaryPath(root))
	if err != nil || !info.Mode().IsRegular() {
		return false
	}
	stamp, err := os.ReadFile(filepath.Join(root, helperSourceStampName))
	if err != nil {
		return false
	}
	return strings.TrimSpace(string(stamp)) == HelperSourceSHA256()
}

// BuildWritebackHelper compiles the embedded Swift helper into root when the
// cached binary is missing or stale, and returns the binary path. compile is
// the swiftc invocation (nil for the real one); the build lock serialises
// concurrent uploader runs.
func BuildWritebackHelper(root string, compile func(source, output string) error) (string, error) {
	if hostOS != "darwin" {
		return "", errors.New("Voice Memos write-back is only available on macOS")
	}
	if compile == nil {
		compile = compileSwiftHelper
	}
	if err := os.MkdirAll(root, 0o755); err != nil {
		return "", err
	}
	lockFile, err := os.OpenFile(filepath.Join(root, helperBuildLockFileName), os.O_CREATE|os.O_RDWR, 0o644)
	if err != nil {
		return "", err
	}
	defer lockFile.Close()
	if err := syscall.Flock(int(lockFile.Fd()), syscall.LOCK_EX); err != nil {
		return "", err
	}
	defer syscall.Flock(int(lockFile.Fd()), syscall.LOCK_UN)
	binary := HelperBinaryPath(root)
	if HelperIsCurrent(root) {
		return binary, nil
	}
	temporary, err := os.MkdirTemp(root, "pdw-voice-memos-writeback-")
	if err != nil {
		return "", err
	}
	defer os.RemoveAll(temporary)
	source := filepath.Join(temporary, "VoiceMemosWriteback.swift")
	if err := os.WriteFile(source, writebackHelperSource, 0o644); err != nil {
		return "", err
	}
	output := filepath.Join(temporary, helperExecutableName)
	if err := compile(source, output); err != nil {
		return "", err
	}
	if err := os.Rename(output, binary); err != nil {
		return "", err
	}
	if err := os.WriteFile(filepath.Join(root, helperSourceStampName), []byte(HelperSourceSHA256()+"\n"), 0o644); err != nil {
		return "", err
	}
	return binary, nil
}

func compileSwiftHelper(source, output string) error {
	xcrun, err := exec.LookPath("xcrun")
	if err != nil {
		return errors.New("Building the Voice Memos write-back helper requires the macOS Command Line Tools")
	}
	cmd := exec.Command(xcrun, "swiftc", "-O", source, "-o", output)
	var stderr, stdout bytes.Buffer
	cmd.Stderr = &stderr
	cmd.Stdout = &stdout
	done := make(chan error, 1)
	if err := cmd.Start(); err != nil {
		return err
	}
	go func() { done <- cmd.Wait() }()
	select {
	case err := <-done:
		if err != nil {
			detail := strings.TrimSpace(stderr.String())
			if detail == "" {
				detail = strings.TrimSpace(stdout.String())
			}
			return fmt.Errorf("Could not compile the Voice Memos write-back helper: %s", detail)
		}
		return nil
	case <-time.After(helperBuildTimeout):
		_ = cmd.Process.Kill()
		return errors.New("Could not compile the Voice Memos write-back helper: swiftc timed out")
	}
}

// helperRequest is the JSON the helper reads on stdin.
type helperRequest struct {
	StorePath string           `json:"store_path"`
	ModelPath string           `json:"model_path"`
	Author    string           `json:"author"`
	DryRun    bool             `json:"dry_run"`
	Items     []RenamePlanItem `json:"items"`
}

// DefaultStoreWriter applies the plan through the Swift helper.
func DefaultStoreWriter(storePath string, items []RenamePlanItem, author string, dryRun bool) ([]WriteResult, error) {
	binary, err := BuildWritebackHelper(DefaultHelperDir(), nil)
	if err != nil {
		return nil, err
	}
	return RunStoreWriterHelper(binary, storePath, items, author, dryRun)
}

// RunStoreWriterHelper invokes a built helper: the model archive is inflated
// here (LoadCachedModelData) and handed over as a temp file, the plan goes in
// on stdin, and the results come back as a JSON array on stdout.
func RunStoreWriterHelper(binary, storePath string, items []RenamePlanItem, author string, dryRun bool) ([]WriteResult, error) {
	if author == "" {
		author = TransactionAuthor
	}
	model, err := LoadCachedModelData(storePath)
	if err != nil {
		return nil, err
	}
	modelFile, err := os.CreateTemp("", "pdw-voice-memos-model-*.plist")
	if err != nil {
		return nil, err
	}
	defer os.Remove(modelFile.Name())
	if _, err := modelFile.Write(model); err != nil {
		modelFile.Close()
		return nil, err
	}
	if err := modelFile.Close(); err != nil {
		return nil, err
	}
	if items == nil {
		items = []RenamePlanItem{}
	}
	request, err := json.Marshal(helperRequest{StorePath: storePath, ModelPath: modelFile.Name(), Author: author, DryRun: dryRun, Items: items})
	if err != nil {
		return nil, err
	}
	cmd := exec.Command(binary)
	cmd.Stdin = bytes.NewReader(request)
	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr
	done := make(chan error, 1)
	if err := cmd.Start(); err != nil {
		return nil, err
	}
	go func() { done <- cmd.Wait() }()
	select {
	case err := <-done:
		if err != nil {
			detail := strings.TrimSpace(stderr.String())
			if detail == "" {
				detail = err.Error()
			}
			return nil, fmt.Errorf("Voice Memos store writer failed: %s", detail)
		}
	case <-time.After(helperRunTimeout):
		_ = cmd.Process.Kill()
		return nil, errors.New("Voice Memos store writer timed out")
	}
	var results []WriteResult
	if err := json.Unmarshal(bytes.TrimSpace(stdout.Bytes()), &results); err != nil {
		return nil, fmt.Errorf("Voice Memos store writer returned invalid output: %w", err)
	}
	if results == nil {
		results = []WriteResult{}
	}
	return results, nil
}
